libs/tree-sitter-wing/grammar.js

```diff
@@ -24,44 +24,54 @@ module.exports = grammar({
 
   word: ($) => $.identifier,
 
+  externals: ($) => [$.AUTOMATIC_SEMICOLON],
+
   precedences: ($) => [
     // Handle ambiguity in case of empty literal: `a = {}`
     // In this case tree-sitter doesn't know if it's a set or a map literal so just assume its a map
     [$.map_literal, $.set_literal],
     [$.json_literal, $.structured_access_expression],
   ],
 
-  conflicts: ($) => [
-    [$.expression, $.custom_type]
-  ],
+  conflicts: ($) => [[$._reference_identifier, $._type_identifier]],
 
   supertypes: ($) => [$.expression, $._literal],
 
   rules: {
     // Basics
     source: ($) => repeat($._statement),
     block: ($) => seq("{", optional(repeat($._statement)), "}"),
+    _semicolon: ($) => choice(";", $.AUTOMATIC_SEMICOLON),
     comment: ($) =>
       token(
         choice(seq("//", /.*/), seq("/*", /[^*]*\*+([^/*][^*]*\*+)*/, "/"))
       ),
 
     // Identifiers
-    reference: ($) => choice($.nested_identifier, $.identifier),
+    reference: ($) =>
+      choice(
+        $.nested_identifier,
+        $._reference_identifier,
+        $.structured_access_expression
+      ),
 
     identifier: ($) => /([A-Za-z_$][A-Za-z_$0-9]*|[A-Z][A-Z0-9_]*)/,
+    _type_identifier: ($) => alias($.identifier, $.type_identifier),
+    _member_identifier: ($) => alias($.identifier, $.member_identifier),
+    _reference_identifier: ($) => alias($.identifier, $.reference_identifier),
 
```

Comment:
Why the distinction between the three types of identifiers? Seems like they are all the same...

Reply:
Initially, I added this because it's handy for debugging grammar conflicts. It makes it clear what rules are used in the grammar because these identifiers are used in specific places. I didn't have to trace the rules like we did on that call.
An additional small benefit is that it makes the grammar much more expressive for the purpose of queries. With that, I could make the tree-sitter highlight query better distinguish between a variable and a property.

afaik it's doesn't hurt to have these, so I figured I would leave it in as a net-positive